addons/web/static/src/core/assets.js

```diff
@@ -0,0 +1,254 @@
+/** @odoo-module **/
+
+const { useEnv, onWillStart } = owl.hooks;
+import { memoize } from "./utils/functions";
+import { browser } from "./browser/browser";
+
+//------------------------------------------------------------------------------
+// Types
+//------------------------------------------------------------------------------
+
+/**
+ * An object describing a bundle to load
+ * @typedef {Object} Bundle
+ * @property {boolean} [templates] whether to load the qweb templates
+ * @property {boolean} [js] whether to load the js from the bundle. Currently
+ *      not implemented.
+ * @property {boolean} [css] whether to load the css from the bundle. Currently
+ *      not implemented.
+ */
+
+/**
+ * An object describing a bundle to load. The keys are bundle names.
+ * @typedef {Object<string, Bundle>} Bundles
+ */
+
+/**
+ * An object describing a loaded bundle
+ * @typedef {Object} LoadedBundle
+ * @property {XMLDocument} templates an XML document containing the owl
+ *      templates defined in that bundle
+ */
+
+/**
+ * An object describing a loaded bundle. The keys are bundle names.
+ * @typedef {Object<string, LoadedBundle>} LoadedBundles
+ */
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+function _loadJS(url) {
+    if (document.querySelector(`script[src="${url}"]`)) {
+        // Already in the DOM and wasn't loaded through this function
+        // Unfortunately there is no way to check whether a script has loaded
+        // or not (which may not be the case for async/defer scripts)
+        // so we assume it is.
+        return Promise.resolve();
+    }
+    const scriptEl = document.createElement("script");
+    scriptEl.type = "text/javascript";
+    scriptEl.src = url;
+    document.head.appendChild(scriptEl);
+    return new Promise((resolve, reject) => {
+        scriptEl.addEventListener("load", resolve);
+        scriptEl.addEventListener("error", reject);
+    });
+}
+/**
+ * Loads the given url inside a script tag.
+ *
+ * @param {string} url the url of the script
+ * @returns {Promise} resolved when the script has been loaded
+ */
+const loadJS = memoize(_loadJS);
```

Comment:
i wonder if it is not better to simply define the function inside here, instead of as a separate function